File: frontend/src/api/models/Condominium.ts
import type { Unit } from './Unit.ts'

export const condominiumCondominiumTypeEnum = {
  BUSINESS: 'BUSINESS',
  RESIDENTIAL: 'RESIDENTIAL',
} as const

export type CondominiumCondominiumTypeEnum = (typeof condominiumCondominiumTypeEnum)[keyof typeof condominiumCondominiumTypeEnum]

export type Condominium = {
  /**
   * @type string | undefined, uuid
   */
  id?: string
  /**
   * @type string | undefined
   */
  name?: string
  /**
   * @type string | undefined
   */
  cnpj?: string
  /**
   * @type string | undefined
   */
  condominiumType?: CondominiumCondominiumTypeEnum
  /**
   * @type string | undefined, date-time
   */
  createdAt?: string
  /**
   * @type string | undefined, date-time
   */
  updatedAt?: string
  /**
   * @type array | undefined
   */
  units?: Unit[]
}

File: backend/src/main/java/br/com/visitsafe/model/condominium/Condominium.java
package br.com.visitsafe.model.condominium;

import br.com.visitsafe.model.enums.CondominiumTypeEnum;
import br.com.visitsafe.model.unit.Unit;
import jakarta.persistence.*;
import lombok.*;

import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "condominiums")
@Inheritance(strategy = InheritanceType.JOINED)
@Data
@NoArgsConstructor
@AllArgsConstructor
public abstract class Condominium {

    @Id
    @GeneratedValue
    private UUID id;

    @Column(nullable = false)
    private String name;

    @Column(nullable = false, unique = true)
    private String cnpj;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, updatable = false)
    private CondominiumTypeEnum condominiumType;

    @Column(nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @Column(nullable = false)
    private OffsetDateTime updatedAt;

    protected Condominium(String name, String cnpj, CondominiumTypeEnum type) {
        this.name = name;
        this.cnpj = cnpj;
        this.condominiumType = type;
    }

    @OneToMany(mappedBy = "condominium",
            cascade = CascadeType.ALL,
            orphanRemoval = true)
    private List<Unit> units = new ArrayList<>();

    @PrePersist
    private void onCreate() {
        createdAt = updatedAt = OffsetDateTime.now();
    }

    @PreUpdate
    private void onUpdate() {
        updatedAt = OffsetDateTime.now();
    }
}
